Declarative Shops

BasicallyIAmFox opened the pull request:
### What is the new feature?
Modernizes NPC shops. #3067.
Replaces `ModNPC.SetupShop` hook with `ModNPC.AddShops` (see Porting Notes)
`ModNPC.ModifyActiveShop` allows manual shop creation in old style.

### Why should this be part of tModLoader?
It could allow checking what items NPC can sell or not, better mod compatibility, and also allows easily insert item before/after X item, as well as adding some conditions to already existing items. Also avoids possible IndexOutOfRangeException, if too many items were added in single shop. Info mods can show how to obtain X item because conditions for them are localized.

### Are there alternative designs?
This PR offers a good compromise between flexibility (via `AbstractNPCShop`) and standardisation. Some aspects are not ideal like the implementation of the 'fixed layout' bartender, but work pretty well overall.

### Sample usage for the new feature
```
public override void AddShops()
{
    var shop = new NPCShop(Type, "Shop"); // You can have different internal names for your shop
    shop.Add<ExampleItem>(); // Adds Example Item in the shop.

    shop.InsertBefore(ModContent.ItemType<ExampleItem>(), ModContent.ItemType<ExampleLifeFruit>()); // Puts Example Life Fruit right before Example Item.

    shop.GetEntry(ModContent.ItemType<ExampleBuffPotion>()).Disable(); // Hides Example Buff Potion from the shop. Will throw an exception if item not found.
    // Safely hides Example Buff Potion from the shop.
    if (shop.TryGetEntry(ModContent.ItemType<ExampleBuffPotion>(), out var entry)) {
        entry.Disable();
    }

    shop.Add(ModContent.ItemType<ExampleSoul>(), NPCShop.Condition.BloodMoon); // Puts an item if its Blood Moon.

    shop.Register();
}
```

### ExampleMod updates
`ExampleNPCShop`, `WeaponWithGrowingDamage`, `ExampleGlobalNPC` were updated.
And of course all town npcs. (`ExamplePerson`, `ExampleTravelingMerchant` and `ExampleBoneMerchant`)

### Porting Notes
`ModPylon.IsPylonForSale(int npcType, Player player, bool isNPCHappyEnough)` -> `ModPylon.GetNPCShopEntry()`
`ModNPC.SetupShop(Chest shop, ref int nextSlot)` -> `ModNPC.ModifyActiveShop()`
`ModNPC.OnChatButtonClicked(bool firstButton, ref bool shop)` -> `ModNPC.OnChatButtonClicked(bool firstButton, ref string shopName)`
`GlobalNPC.SetupShop(int type, Chest shop, ref int nextSlot)` -> `GlobalNPC.ModifyActiveShop(NPC npc, string shopName, Item[] items)`

#### FAQ
- Does this remove the 40 item per shop limit?
  - No, but it should be compatible with a future PR that does.
  - In the mean time, a warning message will be displayed to the user when the shop has too many items to fit.
- How can I randomize my shop items?
  - True randomization would mean contents change every time the shop is opened, and is probably not desirable. Better to use something like moon phases, or a static variable which you change periodically. 
  - Several moon phase related conditions have been added, mostly to support skeleton merchant, but you can make your own conditions with a localization like "Randomly, with 10% chance depending on moon phase" or similar.
- How can I explicitly lay out my shop like DD2 bartender?
  - The only way to do that would be through adding entries and to call `Entry.ReserveSlot()` method afterwards. If conditions are not met, then it will put 'air' item instead of simply not doing anything. Example:
  ```
  shop.Add(
    new NPCShop.Entry(
      ModContent.ItemType<ExampleItem>(),
      Condition.IsMoonFull)
    .ReserveSlot()
  )
  ```
- Can I use this to add items to travelling merchant's shop?
  - Not at this stage. no.
  - Please continue to use `ModifyActiveShop`
    - You can use `TravellingMerchantShop.AddInfoEntry` to add entries to the 'informational' list of the shop, so mods like recipe browser can show how to obtain the item.

BasicallyIAmFox committed:
yay

Solxanich commented:
Internal note: previously #3072

Solxanich commented on patches/tModLoader/Terraria/ModLoader/TMLLootDatabase.cs:
```diff
@@ -0,0 +1,831 @@
+	// can and/or will be used for other tml loots later. (ie pots loot)
+	public class TMLLootDatabase
+	{
+		private Dictionary<int, ChestLoot> npcShopByType = new();
+		private List<ChestLoot.Entry> globalNpcShopEntries = new();
```
proposed is `private Dictionary<string, ChestLoot> npcShopByName = new();`
and when registering for a mod, we need `ModNPC.FullName + "/" + ShopTitle`
For vanilla, let's use Terraria/{NPCName}/ShopTitle

By default, when registering a new Shop, shop title is 'Shop'. Modder can change during setup shop step.

Solxanich commented on patches/tModLoader/Terraria/ModLoader/TMLLootDatabase.cs:
```diff
@@ -0,0 +1,831 @@
+		public List<ChestLoot.Entry> GetGlobalNpcShopEntries() => globalNpcShopEntries;
+
+		public void Initialize() {
+			RegisterMerchant();
+			RegisterArmsDealer();
+			RegisterDryad();
+			RegisterBombGuy();
+			RegisterClothier();
+			RegisterGoblin();
+			RegisterWizard();
+			RegisterMechanic();
+			RegisterSantaClaus();
+			RegisterTruffle();
+			RegisterSteampunker();
+			RegisterDyeTrader();
+			RegisterPartyGirl();
+			RegisterCyborg();
+			RegisterPainter();
+			RegisterWitchDoctor();
+			RegisterPirate();
+			RegisterStylist();
+			RegisterSkeletonMerchant();
+			RegisterBartender();
+			RegisterGolfer();
+			RegisterZoologist();
+			RegisterPrincess();
```
Double check for updates to shops from 1.4.4
For example Painter has a few extra items I believe

Solxanich commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.cs:
```diff
@@ -0,0 +1,321 @@
+		return true;
+	}
+
+	private bool PutAt(int destination, Item item, bool after, params ICondition[] condition) {
+		putCandidates.Add(new(destination, after));
+		putCandidates2.Add(new(item, condition));
+		return true;
+	}
+
+	private bool PutAt(int destination, int item, bool after, params ICondition[] condition) {
+		return PutAt(destination, ContentSamples.ItemsByType[item], after, condition);
+	}
+
+	public bool InsertBefore(int destination, int item, params ICondition[] condition) {
+		return PutAt(destination, item, false, condition);
+	}
+
+	public bool InsertBefore(int destination, Item item, params ICondition[] condition) {
+		return PutAt(destination, item, false, condition);
+	}
+
+	public bool InsertAfter(int destination, int item, params ICondition[] condition) {
+		return PutAt(destination + 1, item, true, condition);
+	}
+
+	public bool InsertAfter(int destination, Item item, params ICondition[] condition) {
+		return PutAt(destination + 1, item, true, condition);
+	}
```
Is Destination referring to the index of the slot or the target Item to be before/after?

If the latter, please rename to targetItem.
If the former, might need some more work

Solxanich commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.cs:
```diff
@@ -0,0 +1,321 @@
+
+	public sealed class Condition : ICondition {
+		#region Common conditions
+		public static readonly Condition Christmas = new(NetworkText.Empty, () => Main.xMas);
+		public static readonly Condition Halloween = new(NetworkText.Empty, () => Main.halloween);
+		public static readonly Condition Hardmode = new(NetworkText.Empty, () => Main.hardMode);
+		public static readonly Condition PreHardmode = new(NetworkText.Empty, () => !Main.hardMode);
+		public static readonly Condition TimeDay = new(NetworkText.Empty, () => Main.dayTime);
+		public static readonly Condition TimeNight = new(NetworkText.Empty, () => !Main.dayTime);
+		public static readonly Condition NotBloodMoon = new(NetworkText.Empty, () => !Main.bloodMoon);
+		public static readonly Condition BloodMoon = new(NetworkText.Empty, () => Main.bloodMoon);
+		public static readonly Condition Eclipse = new(NetworkText.Empty, () => Main.eclipse);
+		public static readonly Condition NotEclipse = new(NetworkText.Empty, () => !Main.eclipse);
+		public static readonly Condition RemixWorld = new(NetworkText.Empty, () => Main.remixWorld);
+		public static readonly Condition TenthAnniversary = new(NetworkText.Empty, () => Main.tenthAnniversaryWorld);
+		public static readonly Condition BirthdayPartyIsUp = new(NetworkText.Empty, () => BirthdayParty.PartyIsUp);
+		public static readonly Condition NightLanternsUp = new(NetworkText.Empty, () => LanternNight.LanternsUp);
+		public static readonly Condition IsMoonFull = new(NetworkText.Empty, () => Main.GetMoonPhase() == MoonPhase.Full);
+		public static readonly Condition IsMoonWaningGibbous = new(NetworkText.Empty, () => Main.GetMoonPhase() == MoonPhase.ThreeQuartersAtLeft);
+		public static readonly Condition IsMoonThirdQuarter = new(NetworkText.Empty, () => Main.GetMoonPhase() == MoonPhase.HalfAtLeft);
+		public static readonly Condition IsMoonWaningCrescent = new(NetworkText.Empty, () => Main.GetMoonPhase() == MoonPhase.QuarterAtLeft);
+		public static readonly Condition IsMoonNew = new(NetworkText.Empty, () => Main.GetMoonPhase() == MoonPhase.Empty);
+		public static readonly Condition IsMoonWaxingCrescent = new(NetworkText.Empty, () => Main.GetMoonPhase() == MoonPhase.QuarterAtRight);
```
We will want to replace the NetworkText.Empty with something meaningful before merge.

Solxanich commented on patches/tModLoader/Terraria/ModLoader/TMLLootDatabase.cs:
```diff
@@ -0,0 +1,831 @@
+			RegisterNpcShop(11, shop);
+		}
+
+		private void RegisterDyeTrader() {
+			ChestLoot shop = new();
+			shop.Add(1037);
+			shop.Add(2874);
+			shop.Add(1120);
+			shop.Add(1969, new ChestLoot.Condition(NetworkText.Empty, () => Main.netMode == NetmodeID.MultiplayerClient));
+			shop.Add(3248, ChestLoot.Condition.Halloween);
+			shop.Add(1741, ChestLoot.Condition.Halloween);
+			shop.Add(2871, ChestLoot.Condition.IsMoonFull);
+			shop.Add(2872, ChestLoot.Condition.IsMoonFull);
+			shop.Add(4663, ChestLoot.Condition.BloodMoon);
+			shop.Add(4662, ChestLoot.Condition.InGraveyard);
+			RegisterNpcShop(12, shop);
+		}
```
Can we get ExamplePerson to be updated with an example like this in Setup Shops?

Just a brief couple of Example items being added with some standard conditions, and setting the shop name as Examplium

Solxanich commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.cs:
```diff
@@ -0,0 +1,321 @@
+		}
+
+		slots = array.Count;
+		if (array.Count < 40) {
+			array.AddRange(Enumerable.Repeat(emptyInstance, 40 - array.Count));
+		}
```
Add additional check for array.Count > 40, and if so, output warning to log and chat that all items could not fit in the shop

Solxanich reviewed (changes requested):
Initial Pass Through.

Great job getting this going, I want to say we are more than half way there.

BasicallyIAmFox committed:
changews

BasicallyIAmFox committed:
Update Chest.cs.patch

BasicallyIAmFox committed:
meow

BasicallyIAmFox committed:
forgot

BasicallyIAmFox committed:
Update TMLLootDatabase.cs

BasicallyIAmFox committed:
Update TMLLootDatabase.cs

BasicallyIAmFox committed:
uh

BasicallyIAmFox committed:
Update ExamplePerson.cs

Solxanich committed:
A series of compilation fixes

Solxanich committed:
Final Compilation fixes

BasicallyIAmFox committed:
Update TMLLootDatabase.cs

BasicallyIAmFox committed:
Update TMLLootDatabase.cs

Solxanich committed:
Large Rework to Get it to Load in to a World for Testing

BasicallyIAmFox committed:
Update TMLLootDatabase.cs

Solxanich committed:
Fixes for Invisible item issues

Solxanich committed:
Re-Add Pylons + Dummyproof EmptyInstance for SetDefaults

BasicallyIAmFox committed:
queen and eol conditions

BasicallyIAmFox committed:
Merge branch '1.4.4_declarativeshops' of https://github.com/BasicallyIAmFox/tModLoader into 1.4.4_declarativeshops

BasicallyIAmFox committed:
painter left.

BasicallyIAmFox committed:
painter

Solxanich committed:
Remove duplicate shop id conversion array + example global PostSetupShop

Solxanich committed:
Global NPC Test InsertAfter

BasicallyIAmFox committed:
pylons

BasicallyIAmFox committed:
trying to fix AddCandidates (№1)

BasicallyIAmFox committed:
.

BasicallyIAmFox committed:
Update ExampleGlobalNPC.cs

BasicallyIAmFox committed:
should we change Entry to be class?

BasicallyIAmFox committed:
Update ExampleBoneMerchant.cs

BasicallyIAmFox committed:
updated parameters for SetupShop

BasicallyIAmFox committed:
Update ExampleGlobalNPC.cs

BasicallyIAmFox committed:
Update ChestLoot.Entry.cs

Solxanich reviewed (approved):
Pulled, tested, seems to be working well.
Code at high level looks good and not seeing any major red flags.

SetupLootboxShop is the only thing somewhat iffy about. I feel like that could return delegate, or return entire array, and allow us to get it out of the SetupShop core. 
I'll leave that for a second opinion

BasicallyIAmFox committed:
removed old SetupShop hook and PylonShopNPC (not needed anymore hopefully)

BasicallyIAmFox committed:
ModifyActiveShop hook

BasicallyIAmFox committed:
Update NPCLoader.cs

BasicallyIAmFox committed:
Update ExampleGlobalNPC.cs

BasicallyIAmFox committed:
example mod fixes and expert/master conditions

BasicallyIAmFox committed:
Update Chest.cs.patch

Chicken-Bones commented on ExampleMod/Common/GlobalItems/WeaponWithGrowingDamage.cs:
```diff
@@ -144,15 +144,24 @@ public class WeaponWithGrowingDamage : GlobalItem
+				Item item = items[i];
+				if (item.type == ItemID.Snowball && item.TryGetGlobalItem(out WeaponWithGrowingDamage weapon)) {
+					weapon.GainExperience(item, 2); // can buy snowballs with 2xp!
+				}
```
This doesn't need to use `ModifyActiveShop`, just need to use the `Item` variant when adding the snowball to the shop.

Chicken-Bones commented on ExampleMod/Common/GlobalNPCs/ExampleNPCShop.cs:
```diff
@@ -7,40 +7,31 @@ namespace ExampleMod.Common.GlobalNPCs
 	{
-		public override void SetupShop(int type, Chest shop, ref int nextSlot) {
+		public override void SetupShop(string shopId, ChestLoot shopContents) {
 			// This example does not use the AppliesToEntity hook, as such, we can handle multiple npcs here by using if statements.
```
Comment no-longer relevant.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/TMLLootDatabase.cs:
```diff
@@ -0,0 +1,1240 @@
+			return null;
+		}
+
+		public static string CalculateShopName(int npcId, string shopName = "Shop")
```
Prefer `GetNPCShopName`

Chicken-Bones commented on ExampleMod/Common/GlobalNPCs/ExampleNPCShop.cs:
```diff
@@ -7,40 +7,31 @@ namespace ExampleMod.Common.GlobalNPCs
+		public override void SetupShop(string shopId, ChestLoot shopContents) {
 			// This example does not use the AppliesToEntity hook, as such, we can handle multiple npcs here by using if statements.
-			if (type == NPCID.Dryad) {
+			if (shopId == TMLLootDatabase.CalculateShopName(NPCID.Dryad, "Shop")) {
```
Lets not pass the `"Shop"` arg as it's default

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.cs:
```diff
@@ -0,0 +1,156 @@
+
+			return putCandidates2.ElementAtOrDefault(index);
+		}
+	}
```
I see no reason to support indexing here.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.cs:
```diff
@@ -0,0 +1,156 @@
+		return PutAt(targetItem, item, true, condition);
+	}
+
+	public ChestLoot Hide(int item) {
+		this[item].Hide();
+		return this;
+	}
```
As above, if people want to hide an entry, they should probably be searching for it, and then they have an `Entry` index they can call. I'd advise against hard-coding vanilla entry indices in mods

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.cs:
```diff
@@ -0,0 +1,156 @@
+	private void AddCandidates(List<Entry> entries) {
+		List<(int nextTo, bool after)> candidates = putCandidates;
+		List<Entry> candidates2 = putCandidates2;
+		candidates.Reverse();
+		candidates2.Reverse();
```
This mutates the original lists... so calling `AddCandidates` multiple times will mess up

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.cs:
```diff
@@ -0,0 +1,156 @@
+		putCandidates.Add(new(destination, after));
+		putCandidates2.Add(new(item, condition));
+		return this;
+	}
```
Please see `Recipe.Ordering` and `SetOrdering` for a good design on how to implement a simple ordering system. 
- Ordering should be specified using `Entry` references, not indices. 
- It's ok to provide a helper which takes `int targetItem` and searches for an entry with that item id, but not required
- Ordering loops should be reported as soon as they're formed

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.cs:
```diff
@@ -0,0 +1,156 @@
+	}
+
+	public Item[] Build(out int slots, bool lastSlotEmpty = true) {
+		List<Item> array = new();
```
Any reason to use a list instead of an `Item[40]`?

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.cs:
```diff
@@ -0,0 +1,156 @@
+		}
+
+		slots = array.Count;
+		if (array.Count > 39) {
+			Main.NewText("Not all items could fit in the shop!", 255, 0, 0);
+			Logging.tML.Warn($"Unable to fit all item in the shop {name}");
+			slots = 39;
```
This doesn't account for `lastSlotEmpty`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.Condition.cs:
```diff
@@ -0,0 +1,115 @@
+
+	public sealed class Condition : ICondition {
+		#region Common conditions
+		public static readonly Condition Christmas = new(NetworkText.FromLiteral("During Christmas"), () => Main.xMas);
```
Need to add localizations for all of these. No `FromLiteral`

Intent is to change `Description` from `string` and `NetworkText` to `LocalizedText` soon

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.Entry.cs:
```diff
@@ -0,0 +1,96 @@
+			return this;
+		}
+
+		public Entry OnFail(int itemId, params ICondition[] condition) {
+			return OnFail(new Entry(ContentSamples.ItemsByType[itemId], condition));
+		}
```
Is `OnFail` necessary? That seems like it would be more appropriate for things that were chance based.
It seems like it would be appropriate for, well, `ChestLoot` or full loot tables, but since this is just conditional visibility I think it might be better to have all conditions be 'top level' (like recipes). Needs discussion

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.Entry.cs:
```diff
@@ -0,0 +1,96 @@
+		internal readonly Item item;
+		private readonly List<ICondition> conditions;
+		private readonly bool askingNicelyToNotAdd = false;
+		private Ref<bool> hide;
```
why `Ref<bool>`?

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.Entry.cs:
```diff
@@ -0,0 +1,96 @@
+	public class Entry {
+		internal readonly Item item;
+		private readonly List<ICondition> conditions;
+		private readonly bool askingNicelyToNotAdd = false;
```
No need for this, just check if `item.IsAir()` in `TryAdd`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.Entry.cs:
```diff
@@ -0,0 +1,96 @@
+			return true;
+		}
+
+		public void TryAdd(List<Item> items) {
```
Generally not advisable to call it `Try` when it doesn't return a `bool` and success doesn't depend on the parameters.

BasicallyIAmFox committed:
changes

BasicallyIAmFox committed:
removed OnFail and OnSuccess

Chicken-Bones commented on ExampleMod/Common/GlobalItems/WeaponWithGrowingDamage.cs:
```diff
@@ -144,15 +144,29 @@ public class WeaponWithGrowingDamage : GlobalItem
+			}
+
+			for (int i = 0; i < items.Length; i++) {
+				Item item = items[i];
```
`foreach (var item in items)`

Chicken-Bones commented on ExampleMod/Common/GlobalItems/WeaponWithGrowingDamage.cs:
```diff
@@ -144,15 +144,29 @@ public class WeaponWithGrowingDamage : GlobalItem
-			if (item.TryGetGlobalItem(out WeaponWithGrowingDamage weapon)) {
-				weapon.GainExperience(item, 2); // can buy snowballs with 2xp!
+			var snowball = new Item();
+			snowball.SetDefaults(ItemID.Snowball);
```
`new Item(ItemID.Snowball)`

Chicken-Bones commented on ExampleMod/Common/GlobalNPCs/ExampleNPCShop.cs:
```diff
@@ -7,40 +7,30 @@ namespace ExampleMod.Common.GlobalNPCs
-				nextSlot++;
+				shopContents.Add(ModContent.ItemType<ExampleMountItem>());
+				shopContents.LastEntry.Item.shopCustomPrice = 2;
+				shopContents.LastEntry.Item.shopSpecialCurrency = CustomCurrencyID.DefenderMedals; // omit this line if shopCustomPrice should be in regular coins.
```
```cs
new Item(ModContent.ItemType<ExampleMountItem>()) {
    shopCustomPrice = 2,
    shopSpecialCurrency = CustomCurrencyID.DefenderMedals
}
```

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModNPC.cs:
```diff
@@ -700,16 +700,24 @@ public virtual void SetChatButtons(ref string button, ref string button2)
-	/// <param name="shop"></param>
-	/// <param name="nextSlot"></param>
-	public virtual void SetupShop(Chest shop, ref int nextSlot)
+	public virtual void SetupShop()
```
This needs to be called `AddShops` akin to `AddRecipes`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModNPC.cs:
```diff
@@ -700,16 +700,24 @@ public virtual void SetChatButtons(ref string button, ref string button2)
+		return null;
+	}
+
+	public virtual void PostSetupShop()
```
There is no use-case for this on `ModNPC`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.cs:
```diff
@@ -0,0 +1,116 @@
+
+namespace Terraria.ModLoader;
+
+public partial class ChestLoot {
```
Needs to be called `NPCShop`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.cs:
```diff
@@ -0,0 +1,116 @@
+		items = new();
+	}
+
+	public void RegisterShop(int npcId, string name = "Shop") {
```
These arguments belong in the constructor, alongside the class rename. Then this can just be `Register`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModNPC.cs:
```diff
@@ -700,16 +700,24 @@ public virtual void SetChatButtons(ref string button, ref string button2)
+	{
+	}
+
+	public virtual void ModifyActiveShop(string shopId, Item[] items)
```
`shopId` -> `shopName`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/GlobalNPC.cs:
```diff
@@ -626,13 +627,19 @@ public virtual void OnChatButtonClicked(NPC npc, bool firstButton)
-	/// <param name="shop"></param>
-	/// <param name="nextSlot"></param>
-	public virtual void SetupShop(int type, Chest shop, ref int nextSlot)
+	public virtual void SetupShop(string shopId, ChestLoot shopContents)
```
Should be called `ModifyShop`

No need for the `shopId` argument, as the name and NPC type of the shop should be obtainable from the `shopContents` parameter (which can just be called `shop`)

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModNPC.cs:
```diff
@@ -700,16 +700,24 @@ public virtual void SetChatButtons(ref string button, ref string button2)
 	/// <param name="firstButton"></param>
 	/// <param name="shop"></param>
-	public virtual void OnChatButtonClicked(bool firstButton, ref bool shop)
+	public virtual void OnChatButtonClicked(bool firstButton, ref string shopName)
```
Need to adjust documentation. Need to be very clear that `shopName` should be set to `"Shop"` for the default shop, and otherwise can be a custom identifier which matches the name of a registered `NPCShop` instance for this NPC.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/GlobalNPC.cs:
```diff
@@ -626,13 +627,19 @@ public virtual void OnChatButtonClicked(NPC npc, bool firstButton)
+	{
+	}
+
+	public virtual void PostSetupShop(string shopId, ChestLoot shopContents)
```
No use case for this. Other loot systems don't have a hook like this.

If, for some reason, there is a need to modify shops after other `GlobalNPC.ModifyShop` hooks have run, `ModSystem.AddRecipes` or `PostAddRecipes` should suffice

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/TMLLootDatabase.cs:
```diff
@@ -0,0 +1,1083 @@
+namespace Terraria.ModLoader;
+
+// can and/or will be used for other tml loots later. (ie pots loot)
+public class TMLLootDatabase
```
Lets keep this specific to shops for now. There's no reason to combine other types of loot tables at this stage, especially since shops and pots and chests all have significant differences.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModNPC.cs:
```diff
@@ -700,16 +700,24 @@ public virtual void SetChatButtons(ref string button, ref string button2)
+	{
+	}
+
+	public virtual List<Item> SetupLootboxShop()
```
It is unclear at this point how to use `SetupLootboxShop`. I understand it's intended to be an access point for travelling merchant style shops, but it doesn't currently have much (if any) advantage over `ModifyActiveShop` for that use case, and could be separated to keep this PR smaller and easier to review

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModContent.cs:
```diff
@@ -337,6 +337,8 @@ internal static void Load(CancellationToken token)
 		ItemSorting.SetupWhiteLists();
 		ItemLoader.ValidateGeodeDropsSet();
 
+		Main.TMLLootDB.Initialize();
```
Should go next to `SetupBestiary`, before `SetupRecipes`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/TMLLootDatabase.cs:
```diff
@@ -0,0 +1,1083 @@
+		npcShopByName.Add(GetNPCShopName(npcId, shopName), chestLoot);
+	}
+
+	public void RegisterGlobalNpcShop(ChestLoot.Entry entry) => globalNpcShopEntries.Add(entry);
```
I am unsure that this is good design. 
On the one hand, being able to see which entries for shops are 'global' could be useful for some mods? 
On the other, this can currently be achieved with a tML `GlobalNPC.ModifyShop` (the load time one, not `ModifyActiveShop`)

With this implementation, it is also unclear _when_ it is valid to register a global shop entry.

Chicken-Bones commented on ExampleMod/Content/NPCs/ExampleGlobalNPC.cs:
```diff
@@ -27,17 +31,54 @@ public class ExampleGlobalNPC : GlobalNPC
+		}
+
+		// PostSetupShop hook is best when it comes to modifying existing items.
+		public override void PostSetupShop(string shopId, ChestLoot shopContents) {
```
This could just be merged with the above method. The rationale for the split only applies if you have a `GlobalNPC` which needs to modify items added by another `GlobalNPC`, in which case you probably have a dependency and mod load order is your friend.

Chicken-Bones commented on ExampleMod/Content/NPCs/ExampleGlobalNPC.cs:
```diff
@@ -27,17 +31,54 @@ public class ExampleGlobalNPC : GlobalNPC
+				return;
+
+			// Creaate a condition for later use.
+			var redPotCondition = new ChestLoot.Condition(NetworkText.FromLiteral("Describe what your condition is!"), () => !ChestLoot.Condition.HappyWindyDay.IsAvailable() || !ChestLoot.Condition.HappyEnough.IsAvailable());
```
Needs to be `NetworkText.FromKey`. Localization is a priority. I'll be changing recipe conditions to use `LocalizedText` soon

Chicken-Bones commented on ExampleMod/Content/NPCs/ExamplePerson.cs:
```diff
@@ -234,64 +234,36 @@ public class ExamplePerson : ModNPC
 				}
 
-				shop = true;
+				shop = "ExampliumShop"; // Name of the shop tab we want to open.
```
It is unclear why this NPC has a non-default shop name. We should make the example include 2 shops if the intent is to showcase this feature properly. Otherwise it's misleading

Chicken-Bones commented on patches/tModLoader/Terraria/Chest.cs.patch:
```diff
@@ -189,6 +195,24 @@
+ 
+ 		int num = 0;
+-		switch (type) {
++		/*switch (type) {
```
put the `/*` on the line above

Chicken-Bones commented on patches/tModLoader/Terraria/Main.cs.patch:
```diff
@@ -1115,6 +1115,35 @@
+ 		ItemDropsDB = itemDropDatabase;
+ 		bestiaryDatabase.Merge(ItemDropsDB);
+ 		bestiaryDatabaseNPCsPopulator.AddDropOverrides(bestiaryDatabase);
++		TMLLootDB = new TMLLootDatabase();
++		TMLLootDB.Initialize();
```
Why have this be an instance, rather than a static class?

Chicken-Bones commented on patches/tModLoader/Terraria/Main.cs.patch:
```diff
@@ -1115,6 +1115,35 @@
++			if (m >= NPCLoader.shopToNPC.Length)
++				continue;
++
++			int npcType = NPCLoader.shopToNPC[m];
++
++			if (m == 25) { // Painter 2 Shop Special Case
++				shop[m].SetupShop(TMLLootDatabase.GetNPCShopName(npcType, "Decor"), m);
++				continue;
++			}
++
++			shop[m].SetupShop(TMLLootDatabase.GetNPCShopName(npcType), m);
```
This lookup code is common in two places, so it should go in `NPCLoader` or the loot database. It should be a function that takes `int shopType` and returns `NPCShop`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.cs:
```diff
@@ -0,0 +1,116 @@
+
+	public IReadOnlyList<Entry> Items => items;
+
+	public Entry LastEntry => items[^1];
+
+	public Entry this[int item] {
+		get {
+			int index = items.FindIndex(x => x.item.type.Equals(item));
+			bool hasInNormal = index != -1;
+			return items[index];
+		}
+	}
```
This should not be a publicly indexable data structure

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.Entry.cs:
```diff
@@ -0,0 +1,112 @@
+			return this;
+		}
+
+		public Entry Hide() {
```
Should be called `Disable`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.Entry.cs:
```diff
@@ -0,0 +1,112 @@
+	public class Entry {
+		internal readonly Item item;
+		private readonly List<ICondition> conditions;
+		private bool hide;
```
Should be `public bool Disabled { get; private set; }`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ChestLoot.Entry.cs:
```diff
@@ -0,0 +1,112 @@
+public partial class ChestLoot {
+	private static Item EmptyInstance => new();
+
+	public class Entry {
```
With `Entry` being public, we need a review of the public API surface, to make sure that all `public` functions are meant to be used by modders, and that there aren't duplicates with the main class.

BasicallyIAmFox committed:
Review changes!

BasicallyIAmFox committed:
Removed this[int] method from NPCShop

BasicallyIAmFox committed:
Removed SetupLootboxShop and renamed TMLLootDatabase to NPCShopDatabase

Chicken-Bones commented on ExampleMod/Common/GlobalItems/WeaponWithGrowingDamage.cs:
```diff
@@ -144,15 +144,27 @@ public class WeaponWithGrowingDamage : GlobalItem
-		public override void SetupShop(int type, Chest shop, ref int nextSlot) {
-			if (type != ModContent.NPCType<ExamplePerson>()) {
+		public override void ModifyShop(NPCShop shop) {
+			if (shop.FullName != NPCShopDatabase.GetNPCShopName(ModContent.NPCType<ExamplePerson>(), "ExampliumShop")) {
```
Should compare `shop.NpcID` and `Name` directly

Chicken-Bones commented on ExampleMod/Common/GlobalItems/WeaponWithGrowingDamage.cs:
```diff
@@ -144,15 +144,27 @@ public class WeaponWithGrowingDamage : GlobalItem
+			shop.Add(snowball);
+		}
+
+		public override void ModifyActiveShop(NPC npc, string shopName, Item[] items) {
+			if (shopName != NPCShopDatabase.GetNPCShopName(ModContent.NPCType<ExamplePerson>(), "ExampliumShop")) {
+				return;
+			}
+
+			foreach (Item item in items) {
+				if (item.type == ItemID.Snowball && item.TryGetGlobalItem(out WeaponWithGrowingDamage weapon)) {
+					weapon.GainExperience(item, 2); // can buy snowballs with 2xp!
+				}
```
This code seems unnecessary? A bad port? You already add snowballs with 2xp to the shop in `ModifyShop`. This would increase their `xp` to 4

Chicken-Bones commented on ExampleMod/Common/GlobalNPCs/ExampleNPCShop.cs:
```diff
@@ -7,40 +7,32 @@ namespace ExampleMod.Common.GlobalNPCs
-			// This example does not use the AppliesToEntity hook, as such, we can handle multiple npcs here by using if statements.
-			if (type == NPCID.Dryad) {
+		public override void ModifyShop(NPCShop shop) {
+			if (shop.FullName == NPCShopDatabase.GetNPCShopName(NPCID.Dryad)) {
```
As above, for NPCs with only one shop, you can just check `shop.NpcID`

Chicken-Bones commented on ExampleMod/Content/NPCs/ExampleGlobalNPC.cs:
```diff
@@ -27,17 +31,53 @@ public class ExampleGlobalNPC : GlobalNPC
+			// Adding new Condition to Blue Flare. Now it will appear just if player carries a Flare Gun in their inventory AND is in Snow biome
+			shop.GetEntry(ItemID.BlueFlare).AddCondition(NPCShop.Condition.InSnowBiome);
+
+			// Creaate a condition for later use.
+			var redPotCondition = new NPCShop.Condition(NetworkText.FromKey("Mods.ExampleMod.ShopConditions.RedPotCondition"), () => !NPCShop.Condition.HappyWindyDay.IsAvailable() || !NPCShop.Condition.HappyEnough.IsAvailable());
```
Should go next to the usage. 
Comment should say 'custom condition, opposite of conditions for `ExampleItem` above'

Chicken-Bones commented on ExampleMod/Content/NPCs/ExamplePerson.cs:
```diff
@@ -234,64 +234,36 @@ public class ExamplePerson : ModNPC
-		// 	// }
-		// }
+		public override void AddShops() {
+			var exampleGunBulletCondition = new NPCShop.Condition(NetworkText.FromLiteral("If its third quarter or waning crescent moon"), () => Main.moonPhase >= 2 && Main.moonPhase < 4);
```
No `FromLiteral`!

Chicken-Bones commented on ExampleMod/Content/NPCs/ExamplePerson.cs:
```diff
@@ -234,64 +234,36 @@ public class ExamplePerson : ModNPC
-		// }
+		public override void AddShops() {
+			var exampleGunBulletCondition = new NPCShop.Condition(NetworkText.FromLiteral("If its third quarter or waning crescent moon"), () => Main.moonPhase >= 2 && Main.moonPhase < 4);
+			var chestLoot = new NPCShop(Type, "ExampliumShop")
```
needs a rename from `chestLoot`

Chicken-Bones commented on ExampleMod/Content/NPCs/ExampleTravelingMerchant.cs:
```diff
@@ -248,22 +248,16 @@ class ExampleTravelingMerchant : ModNPC
-				nextSlot++;
-			}
-		}
+		// TODO: add that method
+		/*public override List<Item> SetupLootboxShop() {
+			return shopItems;
+		}*/
```
Hm? Still need the old code, to fill out the shop the old fashioned way

Chicken-Bones commented on ExampleMod/Localization/en-US.hjson:
```diff
@@ -186,5 +186,10 @@ Mods: {
+
+		#This is used for shop conditions.
+		ShopConditions: {
+			RedPotCondition: "Describe what your condition is!"
```
Proper localization please. `"When not happy enough or not a happy windy day"` or something...

Chicken-Bones commented on patches/tModLoader/Terraria/Chest.cs.patch:
```diff
@@ -189,6 +195,24 @@
+ 		}
+ 
+-		int num = 0;
++		/*int num = 0;
```
Start block comment on line above

Chicken-Bones commented on patches/tModLoader/Terraria/Chest.cs.patch:
```diff
@@ -204,16 +228,44 @@
++		*/
  
++		int num = 0;
++		if (type == 19) // Travelling Merchant, move in to the LootboxShop setup method
```
comment not accurate anymore, remove

Chicken-Bones commented on patches/tModLoader/Terraria/Main.cs.patch:
```diff
@@ -1115,6 +1115,19 @@
+ 			shop[m] = new Chest();
+-			shop[m].SetupShop(m);
++
++			if (m >= NPCLoader.shopToNPC.Length)
++				continue;
++
++			shop[m].SetupShop(NPCShopDatabase.GetNpcShopByIndex(m), m);
```
Is this necessary? Can we just leave `shop` full of blank `new Chest()` values until the player opens one of them, since we rebuild every time anyway?

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/GlobalNPC.cs:
```diff
@@ -626,13 +627,11 @@ public virtual void OnChatButtonClicked(NPC npc, bool firstButton)
-	/// <param name="shop"></param>
-	/// <param name="nextSlot"></param>
-	public virtual void SetupShop(int type, Chest shop, ref int nextSlot)
+	public virtual void ModifyShop(NPCShop shop)
+	{
+	}
+
+	public virtual void ModifyActiveShop(NPC npc, string shopName, Item[] items)
```
Documentation required

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/NPCLoader.cs:
```diff
@@ -1149,19 +1150,28 @@ public static void OnChatButtonClicked(bool firstButton)
 
-	private delegate void DelegateSetupShop(int type, Chest shop, ref int nextSlot);
-	private static HookList HookSetupShop = AddHook<DelegateSetupShop>(g => g.SetupShop);
+	private delegate void DelegateModifyShop(NPCShop shop);
+	private static HookList HookModifyShop = AddHook<DelegateModifyShop>(g => g.ModifyShop);
```
Should go below `AddShops` next to `ModifyShop`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/NPCShop.cs:
```diff
@@ -0,0 +1,125 @@
+	public string Name => name;
+	public string FullName => NPCShopDatabase.GetNPCShopName(NpcID, Name);
+
+	public Entry this[int item] {
+		get {
+			int index = items.FindIndex(x => x.item.type.Equals(item));
+			bool hasInNormal = index != -1;
+			return items[index];
+		}
+	}
```
Please remove, since we now have `GetEntry`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/NPCShop.cs:
```diff
@@ -0,0 +1,125 @@
+		return Build(out _);
+	}
+
+	public Item[] Build(out int slots) {
```
Is there a reason not to pass in the `Item[]` here? I don't like that this method is hard-coded to 40 slots with a lot of bits and pieces

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/NPCShop.cs:
```diff
@@ -0,0 +1,125 @@
+			if (group.Disabled || !group.ConditionsMet()) {
+				continue;
+			}
+			array.Add(group.Item);
```
could fill into the array directly, and break when hitting 39 items

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/NPCShop.cs:
```diff
@@ -0,0 +1,125 @@
+
+		slots = array.Count;
+		int limit = 39;
+		if (array.Count > limit) {
+			Main.NewText("Not all items could fit in the shop!", 255, 0, 0);
+			Logging.tML.Warn($"Unable to fit all item in the shop {name}");
+			slots = limit;
+		}
```
This doesn't belong in here. Not nice to have methods which seem 'pure' on the outside, to be calling game functions. Better to just have a `bool` return, or an `out bool overflow`

BasicallyIAmFox committed:
Funny changes! (too many to put them all in here)

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/NPCShopDatabase.cs:
```diff
@@ -0,0 +1,1091 @@
+		}).ToDictionary(x => x.Key, x => x.Value);
+	}
+
+	public static string GetNpcShopByIndex(int index)
```
This should be called `GetVanillaShop`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/NPCShopDatabase.cs:
```diff
@@ -0,0 +1,1091 @@
+			return split[0] == (npcId < NPCID.Count ? "Terraria" : NPCLoader.GetNPC(npcId).Mod.Name)
+			&& split[1] == (npcId < NPCID.Count ? NPCID.Search.GetName(npcId) : NPCLoader.GetNPC(npcId).Name);
+		}).ToDictionary(x => x.Key, x => x.Value);
+	}
```
No need for this method, can just allow people to enumerate all shops and then they can check `NPCType`

BasicallyIAmFox committed:
OrderLast, probably fixed SortBeforeAfter and added NoPylons dictionary

BasicallyIAmFox committed:
readded PylonShopNPC, made sorting run after recipes

BasicallyIAmFox committed:
Merge branch '1.4.4' of https://github.com/tModLoader/tModLoader into 1.4.4_declarativeshops

BasicallyIAmFox committed:
guh

BasicallyIAmFox committed:
Dread

BasicallyIAmFox committed:
Mechanic is NOT Merchant!

BasicallyIAmFox committed:
Accessory Slots are evil

BasicallyIAmFox committed:
Moved DrawAccSlots call to be after shop rendering

BasicallyIAmFox committed:
Added missing entries in the shops

BasicallyIAmFox committed:
some documentation and tmodportertests (successful

BasicallyIAmFox committed:
unbroke traveling merchant

BasicallyIAmFox committed:
SetupShop is now ModifyActiveShop. BARTENDER

Chicken-Bones committed:
Merge remote-tracking branch 'origin/1.4.4' into 1.4.4_declarativeshops

Chicken-Bones committed:
Merge branch '1.4.4' into 1.4.4_declarativeshops

Chicken-Bones committed:
Fix tModPorter tests, SetupShop -> ModifyActiveShop

Chicken-Bones committed:
Better documentation and consistent parameter naming

Chicken-Bones committed:
ExampleMod localization auto-export update

Chicken-Bones committed:
Fix typo, Reverse -> Reserve

Chicken-Bones committed:
NPCShop.items -> entries

Chicken-Bones committed:
Improve NPCShop.Build code quality/performance

Chicken-Bones committed:
Make NPCShop.Entry ordering calls more user friendly (like Recipe)

Chicken-Bones committed:
Process NPCShop.Entry.OrdersLast before actual ordering

Chicken-Bones committed:
Refactor shop sorting into its own method. Simplify

Chicken-Bones committed:
Fix conditions in ExampleGlobalNPC

Chicken-Bones committed:
Fix NPCShop.SortBeforeAfter so it actually sorts

Chicken-Bones committed:
Move SortBeforeAfter into NPCShop (not entry)

Chicken-Bones committed:
Move bartender shop setup back into BartenderShopNPC

Chicken-Bones committed:
Simplify iteration in for ModifyShop

Chicken-Bones committed:
Improve code formatting of Bartender shop

Chicken-Bones committed:
Remove NPCShop. qualifiers everywhere in NPCShopDatabase

BasicallyIAmFox committed:
Added Entry.OnShopOpened and made NPCShop.Build to clone items rather than directly passing them.

BasicallyIAmFox committed:
Now Entry.OnShopOpen actually gets called whenever item is added to shop.

Chicken-Bones committed:
Use a shop name for the travelling merchant's shop, for forwards compat

Chicken-Bones committed:
Minor compile fixes and refactors

Chicken-Bones committed:
Simplify SetupShop, always has an NPC reference. Use null for empty slots

Chicken-Bones committed:
Merge branch '1.4.4' into 1.4.4_declarativeshops

Chicken-Bones committed:
Add NPCShop.FillLastSlot

Chicken-Bones committed:
Fix PostBuyItem being called twice

Chicken-Bones committed:
More naming cleanup and property refactors

Chicken-Bones committed:
Cleanup Insert/Add overloads, add InsertBefore/After(Entry)

Chicken-Bones committed:
Fix brace formatting style

Chicken-Bones committed:
Move NPCLoader shopToNpc into NPCShopDatabase

Chicken-Bones committed:
doc fix

Chicken-Bones committed:
Order pylons last

BasicallyIAmFox committed:
Added condition for shimmered npcs. More in desc

- Added an example where item appears just if NPC is shimmered.
- Example where if NPC is shimmered, then all prices are reduces by 50%
- Moved shop examples from ExampleGlobalNPC to ExampleNPCShop
- :person_with_blond_hair:

Chicken-Bones committed:
Cleanup example shop conditions

Chicken-Bones committed:
Slight consistency/readability improvements for some moon phase conditions

Chicken-Bones committed:
Use generic Add method in ExampleMod where possible

Chicken-Bones committed:
Add Player.TalkNPC

Chicken-Bones committed:
Fix wrong pylon conditions

Chicken-Bones committed:
Improve pylon condition localisations

Chicken-Bones committed:
Shop database cleanup, some condition fixes in there

Chicken-Bones committed:
Keep vanilla setup shop method for comparison

Chicken-Bones committed:
Fix NoPylons exclusion, meant to be travelling merchant

Chicken-Bones committed:
Only make pylon entries once

Chicken-Bones committed:
Add automatic shop testing code

Chicken-Bones committed:
Fix several items, conditions and ordering issues in NPCShopDatabase

Chicken-Bones committed:
Change ModPylon to use the new NPCShop.Entry system

Chicken-Bones committed:
Improve localizations

Chicken-Bones committed:
Further improve localizations

Chicken-Bones committed:
Remaining localization key reference issues and naming tweaks

Chicken-Bones committed:
Remove condition parameters from ModPylon.GetNPCShopEntry

BasicallyIAmFox committed:
Fixed order of items in shops.

BasicallyIAmFox committed:
More condition fixes as well as NoPylons now used

BasicallyIAmFox committed:
Fixed compilation error

BasicallyIAmFox committed:
Made pylon entries list to clear on unload (incase it doesn't already)

BasicallyIAmFox committed:
Now using TryGetValue to avoid KeyNotFound exception

Chicken-Bones committed:
Change NoPylons to be a Set

Chicken-Bones committed:
Fix wrong condition for RoyalRomance

Chicken-Bones committed:
Implement pylons for bartender with hack that matches vanilla

Chicken-Bones committed:
Make shop test code ignore position rounding issues

Chicken-Bones committed:
Add localizations for DownedQueenSlime condition

Chicken-Bones committed:
Make shop overflow message less angry, and localize it

Chicken-Bones committed:
Fix bartender pylon selling logic when > 1 pylon sold

Chicken-Bones committed:
Revert the attempts to avoid condition quirks in shop tests

Chicken-Bones committed:
Fix corruption and crimson localizations being reversed

Chicken-Bones committed:
Use >= for golf score condition

Chicken-Bones committed:
Add special steampunker outfit condition

Chicken-Bones committed:
Merge remote-tracking branch 'origin/1.4.4' into 1.4.4_declarativeshops

Chicken-Bones committed:
Disable testing code

Chicken-Bones committed:
Merge remote-tracking branch 'origin/1.4.4' into 1.4.4_declarativeshops

Chicken-Bones committed:
Merge branch '1.4.4' into 1.4.4_declarativeshops

Chicken-Bones committed:
Call GetOrRegister when making conditions via the string ctor overload

Chicken-Bones committed:
Merge NPCShop.Condition into Condition

Chicken-Bones committed:
Slightly more consistent condition localizations

Chicken-Bones committed:
Fix SetupTravelShop hook

Chicken-Bones committed:
Add AbstractNPCShop and info entries for Travelling Merchant

Chicken-Bones committed:
Rename Build -> FillShop. Add ICollection variant for future compat

Chicken-Bones committed:
Rename AbstractNPCShop.IShopEntry to Entry

Chicken-Bones committed:
Better ExampleTravelingMerchant showcasing new shop system

Chicken-Bones committed:
Fix bug with ExampleTravelingMerchant
